File: consolidado/tools/ghidra_scripts/TraceIvOffsetFormula.py
# Ghidra script: Trace IV offset and provider structure fields for GEOM_CRYPT and g_part
#
# Usage:
#   analyzeHeadless <project_loc> <project> \
#       -process kmem_dump_1252.bin -noanalysis \
#       -postScript TraceIvOffsetFormula.py \
#       -scriptPath /mnt/t/downloads/PS4/linux_project/consolidado/tools/ghidra_scripts \
#       -readOnly

from ghidra.app.decompiler import DecompInterface, DecompileOptions
from ghidra.util.task import ConsoleTaskMonitor

print("=== TraceIvOffsetFormula.py starting ===")

options = DecompileOptions()
decomp = DecompInterface()
decomp.setOptions(options)
decomp.openProgram(currentProgram)

fm = currentProgram.getFunctionManager()
addr_factory = currentProgram.getAddressFactory().getDefaultAddressSpace()

def get_func_at(va_hex):
    addr = addr_factory.getAddress(va_hex)
    f = fm.getFunctionAt(addr)
    if f is None:
        # Try to find function containing address
        f = fm.getFunctionContaining(addr)
    return f

# Targets of interest:
# 1. 0xffffffffdc9a40d0 (g_crypt_create_provider)
# 2. 0xffffffffdc9a1ce0 (g_crypt_taste)
# 3. 0xffffffffdc8dabae (g_part_gpt)
# 4. 0xffffffffdc8dab50 (g_part_gpt_create_provider)

target_addrs = [0xffffffffdc9a40d0, 0xffffffffdc9a1ce0, 0xffffffffdc8dabae, 0xffffffffdc8dab50]

for va in target_addrs:
    f = get_func_at(va)
    if f:
        print("Found function %s at 0x%x" % (f.getName(), va))
        res = decomp.decompileFunction(f, 30, ConsoleTaskMonitor())
        if res and res.decompiledFunction:
            ccode = res.decompiledFunction.getC()
            print("--- C Code for 0x%x (%s) ---" % (va, f.getName()))
            lines = ccode.splitlines()
            for idx, line in enumerate(lines[:120]):
                print("%3d: %s" % (idx + 1, line))
            print("--- End C Code ---")
        else:
            print("Failed to decompile 0x%x" % va)
    else:
        print("No function found at 0x%x" % va)

print("=== TraceIvOffsetFormula.py finished ===")

File: consolidado/tools/ghidra_scripts/TraceIvOffsetFormula.java
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.*;
import ghidra.program.model.address.*;
import ghidra.program.model.listing.*;

public class TraceIvOffsetFormula extends GhidraScript {
    @Override
    public void run() throws Exception {
        println("=== TraceIvOffsetFormula.java starting ===");
        DecompInterface decomp = new DecompInterface();
        decomp.openProgram(currentProgram);

        long[] addrs = {
            0xffffffffdc9a40d0L, // g_crypt_create_provider
            0xffffffffdc9a1ce0L, // g_crypt_taste
            0xffffffffdc8dabaeL, // g_part_gpt
            0xffffffffdc8dab50L, // g_part_gpt_create_provider
            0xffffffffdc8d9e30L  // g_part_gpt_taste / init
        };

        AddressFactory af = currentProgram.getAddressFactory();
        FunctionManager fm = currentProgram.getFunctionManager();

        for (long va : addrs) {
            Address addr = af.getDefaultAddressSpace().getAddress(va);
            Function f = fm.getFunctionAt(addr);
            if (f == null) {
                f = fm.getFunctionContaining(addr);
            }
            if (f != null) {
                println("Found function: " + f.getName() + " at " + f.getEntryPoint());
                DecompileResults res = decomp.decompileFunction(f, 30, monitor);
                if (res != null && res.getDecompiledFunction() != null) {
                    String code = res.getDecompiledFunction().getC();
                    println("--- C Code for 0x" + Long.toHexString(va) + " (" + f.getName() + ") ---");
                    println(code);
                    println("--- End C Code ---");
                }
            } else {
                println("No function found at 0x" + Long.toHexString(va));
            }
        }
        println("=== TraceIvOffsetFormula.java finished ===");
    }
}
